thunar/thunar-location-bar.c

```diff
@@ -237,34 +265,46 @@ static GtkWidget *
 thunar_location_bar_install_widget (ThunarLocationBar    *bar,
                                     GType                 type)
 {
-  GtkWidget *widget, *child;
+  GtkWidget *installedWidget, *child;
 
   /* check if the the right type is already installed */
   if ((child = gtk_bin_get_child (GTK_BIN (bar))) && G_TYPE_CHECK_INSTANCE_TYPE (child, type))
     return child;
 
-  widget = gtk_widget_new (type, "current-directory", bar->current_directory, NULL);
-
-  /* forward signals */
-  g_signal_connect_swapped (widget, "change-directory", G_CALLBACK (thunar_navigator_change_directory), THUNAR_NAVIGATOR (bar));
-  g_signal_connect_swapped (widget, "open-new-tab", G_CALLBACK (thunar_navigator_open_new_tab), THUNAR_NAVIGATOR (bar));
-
-  if (type == THUNAR_TYPE_LOCATION_ENTRY)
+  if( type == THUNAR_TYPE_LOCATION_ENTRY )
     {
-      g_signal_connect_swapped (widget, "reload-requested", G_CALLBACK (thunar_location_bar_reload_requested), bar);
+    if( bar->locationEntry == NULL )
+      {
+      bar->locationEntry = gtk_widget_new (THUNAR_TYPE_LOCATION_ENTRY, "current-directory", NULL, NULL);
+      g_object_ref (bar->locationEntry);
+      g_signal_connect_swapped (bar->locationEntry, "reload-requested", G_CALLBACK (thunar_location_bar_reload_requested), bar);
+      g_signal_connect_swapped (bar->locationEntry, "change-directory", G_CALLBACK (thunar_navigator_change_directory), THUNAR_NAVIGATOR (bar));
+      g_signal_connect_swapped (bar->locationEntry, "open-new-tab", G_CALLBACK (thunar_navigator_open_new_tab), THUNAR_NAVIGATOR (bar));
+      }
+    installedWidget = bar->locationEntry;
     }
-  else if (type == THUNAR_TYPE_LOCATION_BUTTONS)
+  else
     {
-      g_signal_connect_swapped (widget, "entry-requested", G_CALLBACK (thunar_location_bar_request_entry), bar);
+    if( bar->locationButtons == NULL )
+      {
+      bar->locationButtons = gtk_widget_new (THUNAR_TYPE_LOCATION_BUTTONS, "current-directory", NULL, NULL);
+      g_object_ref (bar->locationButtons);
+      g_signal_connect_swapped (bar->locationButtons, "entry-requested", G_CALLBACK (thunar_location_bar_request_entry), bar);
+      g_signal_connect_swapped (bar->locationButtons, "change-directory", G_CALLBACK (thunar_navigator_change_directory), THUNAR_NAVIGATOR (bar));
+      g_signal_connect_swapped (bar->locationButtons, "open-new-tab", G_CALLBACK (thunar_navigator_open_new_tab), THUNAR_NAVIGATOR (bar));
+      }
+    installedWidget = bar->locationButtons;
     }
 
+  thunar_navigator_set_current_directory (THUNAR_NAVIGATOR (installedWidget), bar->current_directory);
+
   if ((child = gtk_bin_get_child (GTK_BIN (bar))))
-    gtk_widget_destroy (child);
+    gtk_container_remove (bar,child);
```

Comment:
Introduces warning, please use the GTK_CONTAINER macro. (Don't forget to put a space after the comma)